lib/Runtime/Provisioner/Provisioner.cpp

```diff
@@ -396,6 +428,51 @@ Error Provisioner::provision(DAGListTy &networks, Module &module,
       }
     }
   }
+
+  // If a deferredWeightLoader is provided, create a deferredWeightLoader and
+  // load deferred weights.
+  if (cctx.deferredWeightLoader) {
+
+    auto loader = cctx.deferredWeightLoader;
+    // Load the first weight.
+    RETURN_IF_ERR(loader->loadNextWeight());
+    std::string weightName = loader->getName();
+    // Load weights while there are weights to be loaded.
+    while (weightName != "") {
+      auto PH = module.getPlaceholderByName(weightName);
+      if (!PH) {
+        return MAKE_ERR(
+            ErrorValue::ErrorCode::RUNTIME_ERROR,
+            llvm::formatv(
+                "Error loading deferred weight. Name: {0} not found in module.",
+                weightName)
+                .str());
+      }
+      // Transfer weight to all devices needed.
+      for (const auto &device : placeholderToDevicManager[PH]) {
+        std::promise<Error> transferPromise;
```

Comment:
This breaks Visual Studio build.
Not sure what's the best way to fix it. 
Adding std::promise as a friend doesn't seem to work. It produces other errors on the friend line.
Removing protected, just to try, leads to other errors. I think because constructor is explicit. Both of which were done for a reason.

include\future(226): error C2248: 'glow::detail::GlowError::GlowError': cannot access protected member declared in class 'glow::detail::GlowError'
25>C:\projects\glowExternal\glow\include\glow/Support/Error.h(384): note: see declaration of 'glow::detail::GlowError::GlowError'
25>C:\projects\glowExternal\glow\include\glow/Support/Error.h(341): note: see declaration of 'glow::detail::GlowError'

@jackm321 @gcatron  any ideas?

Reply:
@ayermolo what version of visual studio are you building with?

Reply:
@gcatron Visual Studio 2017 15.9.17.

Reply:
@ayermolo sorry for the delay getting back to you, can you try changing the promise to void and capturing a reference to the promise in the lambda function? Like we do in ExecutionEngine::runInternal? If that compiles correctly with VisualStudio we can change to that.

Reply:
@gcatron No problem, we all busy with internal stuff. :)
Something like this?
```
std::promise<void> transferPromise;
        Error transferError = Error::empty();
        auto done = transferPromise.get_future();
        devices_[device]->transferStaticPlaceholderToDevice(
            PH, weight, [&transferPromise, &transferError](Error err) {
              transferError = std::move(err);
              transferPromise.set_value();
            });
        RETURN_IF_ERR(std::move(transferError));
```

It passes. I also made change in DeviceManagerTest, and it builds and passes.
Alternative is this MR I did:
https://github.com/pytorch/glow/pull/3926
On plus side it solves the issue at the "root" so if down the road someone uses current way it is done it won't break build again. On downside the affect of the change has higher scope.
Right now it's showing OpenCL failing, but I think it's unrelated. I have seen QuantizationTest fail randomly locally for a while now.

Reply:
Exactly like that, that being said your changes in #3926 look fine to me too, I'd like to get feedback from @jackm321 before landing though in case there was a specific reason it's that way.